azure/services/scalesets/spec.go

```diff
@@ -0,0 +1,522 @@
+			vmss.VirtualMachineScaleSetProperties.AdditionalCapabilities = &compute.AdditionalCapabilities{
+				UltraSSDEnabled: pointer.Bool(true),
+			}
+		}
+	}
+
+	// Set Additional Capabilities if any is present on the spec.
+	if s.AdditionalCapabilities != nil {
+		// Set UltraSSDEnabled if a specific value is set on the spec for it.
+		if s.AdditionalCapabilities.UltraSSDEnabled != nil {
+			vmss.AdditionalCapabilities.UltraSSDEnabled = s.AdditionalCapabilities.UltraSSDEnabled
+		}
+	}
+
+	if s.TerminateNotificationTimeout != nil {
+		vmss.VirtualMachineScaleSetProperties.VirtualMachineProfile.ScheduledEventsProfile = &compute.ScheduledEventsProfile{
+			TerminateNotificationProfile: &compute.TerminateNotificationProfile{
+				NotBeforeTimeout: pointer.String(fmt.Sprintf("PT%dM", *s.TerminateNotificationTimeout)),
+				Enable:           pointer.Bool(true),
+			},
+		}
+	}
+
+	tags := infrav1.Build(infrav1.BuildParams{
+		ClusterName: s.ClusterName,
+		Lifecycle:   infrav1.ResourceLifecycleOwned,
+		Name:        pointer.String(s.Name),
+		Role:        pointer.String(infrav1.Node),
+		Additional:  s.AdditionalTags,
+	})
+
+	vmss.Tags = converters.TagsToMap(tags)
+	return vmss, nil
+}
+
+func hasModelModifyingDifferences(infraVMSS *azure.VMSS, vmss compute.VirtualMachineScaleSet) bool {
+	other := converters.SDKToVMSS(vmss, []compute.VirtualMachineScaleSetVM{})
+	return infraVMSS.HasModelChanges(*other)
+}
+
+func (s *ScaleSetSpec) generateExtensions(ctx context.Context) ([]compute.VirtualMachineScaleSetExtension, error) {
+	extensions := make([]compute.VirtualMachineScaleSetExtension, len(s.VMSSExtensionSpecs))
+	for i, extensionSpec := range s.VMSSExtensionSpecs {
+		extensionSpec := extensionSpec
+		parameters, err := extensionSpec.Parameters(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		vmssextension, ok := parameters.(compute.VirtualMachineScaleSetExtension)
+		if !ok {
+			return nil, errors.Errorf("%T is not a compute.VirtualMachineScaleSetExtension", parameters)
+		}
+		extensions[i] = vmssextension
+	}
+
+	return extensions, nil
+}
+
+func (s *ScaleSetSpec) getVirtualMachineScaleSetNetworkConfiguration() *[]compute.VirtualMachineScaleSetNetworkConfiguration {
+	var backendAddressPools []compute.SubResource
+	if s.PublicLBName != "" {
+		if s.PublicLBAddressPoolName != "" {
+			backendAddressPools = append(backendAddressPools,
+				compute.SubResource{
+					ID: pointer.String(azure.AddressPoolID(s.SubscriptionID, s.ResourceGroup, s.PublicLBName, s.PublicLBAddressPoolName)),
+				})
+		}
+	}
+	nicConfigs := []compute.VirtualMachineScaleSetNetworkConfiguration{}
+	for i, n := range s.NetworkInterfaces {
+		nicConfig := compute.VirtualMachineScaleSetNetworkConfiguration{}
+		nicConfig.VirtualMachineScaleSetNetworkConfigurationProperties = &compute.VirtualMachineScaleSetNetworkConfigurationProperties{}
+		nicConfig.Name = pointer.String(s.Name + "-nic-" + strconv.Itoa(i))
+		nicConfig.EnableIPForwarding = pointer.Bool(true)
```

Comment:
Do we want to leave it to what user specifies instead?

Reply:
see https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3224 / https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3243/files for a similar discussion - IIRC at this point it's not configurable and needs to be set to true otherwise breaks networking.